<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<!--
########################################
#  Copyright (c)  2007-2008 Elisha Roberson and Jonathan Pevsner.
#                 All Rights Reserved.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS "AS IS" AND ANY 
#  EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNERS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE IS FREE FOR PERSONAL OR ACADEMIC
#  USE. THE SOFTWARE MAY NOT BE USED COMMERCIALLY WITHOUT THE EXPRESS, WRITTEN
#  PERMISSION OF THE COPYRIGHT HOLDERS. ALL ACTIONS OR PROCEEDINGS RELATED TO 
#  THIS SOFTWARE SHALL BE TRIED EXCLUSIVELY IN THE STATE AND FEDERAL COURTS 
#  LOCATED IN THE COUNTY OF BALTIMORE CITY, MARYLAND. USE OF THIS SOFTWARE
#  IMPLIES ACCEPTANCE OF THIS CONTRACT.
########################################
-->
<html>

<head>
<title>SNPduoWeb at the Pevsnerlab -- Tutorial</title>
<link rel="stylesheet" type="text/css" href="SNPduo.css">
<meta name="author" content="Elisha Roberson">
<meta name="description" content="A tutorial explaining the preparation of data for SNPduoWeb and how to run the tool">
<meta name="keywords" content="SNPduo,SNP,tutorial,how to,guide,Affymetrix,Illumina,Hap Map,custom,data,export,upload,options">
</head>

<body>
<!-- <a name="top"></a> -->
<!-- ################################  Insert header elements similar to the rest of the site  ################################ -->
<table class="headertables"> 
<tr>
<td>
<img src="images/SNPduo_titlegraphic.jpg" width="600" height="90" alt="Pevsner Lab Header Picture">
</td>
</tr>
</table>

<table class="headertables">
<tr>
<td class="menu">
<a href="/index.html">Home</a> | 
<a href="/dragon.htm">DRAGON</a> | 
<a href="/snomad.htm">SNOMAD</a> | 
<a href="/snpscan.htm">SNPscan</a> | 
<a href="/SNPtrio.htm">SNPtrio</a> | 
<a href="/bioinformatics/bioinf.htm">Bioinformatics</a> | 
<a href="/bioinfo_course.htm">Bioinformatics Course</a> | 
<a href="/microarraydata/index.html">Microarrays</a> | 
<a href="/lead.htm">Lead Poisoning</a><br>
<a href="/autism_rett.htm">Autism and Rett Syndrome</a> | 
<a href="/index_ds.htm">Down Syndrome</a> | 
<a href="http://www.davinciandthebrain.org">Leonardo da Vinci</a> | 
<a href="/publications.htm">Publications</a> | 
<a href="/contact.htm">Lab Personnel &amp; Contact Info</a>
</td>
</tr>
</table>
<!-- ################################  End header elements  ################################ -->
<p><a name="top"></a></p>

<table width="100%" border="0" cellpadding="0" cellspacing="0">
<tr class="rowcommon cell2">

<td>
<blockquote>
<table border="0" cellpadding="6" cellspacing="3">

<tr class="rowcommon">
<td class="cell1">Tutorial</td>
<td class="cell2">A tutorial for first-time SNPduoWeb users. Covers the analysis of Affymetrix, Illumina, HapMap, and custom format data</td>
</tr>
</table>
</blockquote>
</td>

</tr>
</table>

<hr>
<p>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Each of the exported data types that SNPduoWeb supports have slightly different formats. SNPduoWeb has been designed to handle the differences in these formats, but the data must be exported in a way SNPduoWeb can understand. Please select a tutorial to learn how to export data and run SNPduo.
<br><br>

Jump to a tutorial<br><br>
<a href="#Affymetrix">Affymetrix CNAT 4.0 Data Export</a><br>
<a href="#Illumina">Illumina BeadStudio Data Export</a><br>
<a href="#HapMap">HapMap Downloaded Data Export</a><br>
<a href="#Custom">Custom Data Formats</a><br>
<a href="#SNPduo">Running SNPduo</a><br>
<a href="#Tips">Tips and Hints</a>
</p>

<hr>

<!-- Affy data analysis -->
<p>
<a name="Affymetrix"><strong>Affymetrix CNAT 4.0 Data</strong></a>
<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Affymetrix data files are typically analyzed using the GeneChip Genotyping Analysis Software. This tutorial was written based on version 4.1.0.26, but should be applicable to older versions as well.


<blockquote>
	<p>
	1. Open the Affymetrix GType software
	<br><br>
	2. In the data source panel choose "Analysis Results"
	</p>
	
	<blockquote>
	<p>
	<img src="images/affy_analysisresults.jpg" width="251" height="186" alt="Affy Analysis Results">
	</p>
	</blockquote>
	
	<p>
	3. Double click on the appropriate "*.chp" files for the individuals you wish to analyze
	<br><br>
	
	4. Several columns are displayed for the data you have chosen. SNPduoWeb will run faster if less extraneous data is displayed. Right click in the column headers for data <strong>other</strong> than Chromosome, Physical Position, and the genotype calls for the individuals of interest, and click the "Hide Column" option
	</p>
	
	<blockquote>
	<p>
	<img src="images/affy_hide.jpg" width="444" height="180" alt="Affy Hide Columns">
	</p>
	</blockquote>
	
	<p>
	5. Click the Export button from the toolbar
	</p>
	
	<blockquote>
	<p>
	<img src="images/affy_export.jpg" width="529" height="63" alt="Affy Export Data">
	</p>
	</blockquote>
	
	<p>
	6. Choose an appropriate file name and save location. Be sure the "Export All" option is clicked and export the data
	</p>
	
	<blockquote>
	<p>
	<img src="images/affy_save.jpg" width="428" height="122" alt="Affy Save File">
	</p>
	</blockquote>
	
	<p>
	7. Proceed with using the SNPduoWeb tool as described <a href="#SNPduo">below</a>
	</p>
</blockquote>

<p>
<a href="#top">top</a>
</p>

<hr>

<!-- Illumina data analysis -->

<p>
<a  name="Illumina"><strong>Illumina - BeadStudio</strong></a>
<br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;For SNPs genotyped using Illumina data the standard analysis software is BeadStudio. This portion of the tutorial will cover how to export data from BeadStudio for use with SNPduo.
</p>


<blockquote>
	<p>
	1. Open your data file (typically *.bsc) with BeadStudio
	<br><br>
	
	2. Make sure you are in the Full Data Table
	<br><br>
	
	3. Using the Column Chooser be sure you have at least the following selected:
	</p>
	
	<blockquote>
	<p>
	<img src="images/illumina_columnchooser.jpg" width="471" height="77" alt="Illumina Column Chooser">
	</p>
	</blockquote>
	
	<p>
	<strong>From "Displayed Columns" show</strong>
	</p>
	<ul>
		<li>Chr</li>
		<li>Position</li>
		<li>The individuals of interest</li>
	</ul>
	
	<p>
	<strong>From "Displayed Subcolumns" show</strong>
	</p>
	<ul>
		<li>GType</li>
	</ul>
	
	<p>
	4. Press the "Export Displayed Data to File" Button. Choose a save location and name. Export the data
	</p>
	
	<blockquote>
	<p>
	<img src="images/illumina_export.jpg" width="483" height="78" alt="Illumina Export Data">
	</p>
	</blockquote>
	
	<p>
	5. SNPduoWeb labels plots based on the column headers for the two compared individuals. To adjust the headers on a BeadStudio file open the exported file in a text editor. Change the genotype headers to whatever you wish to be displayed in the title of each individual's genotype plot and the overall title.
	<br><br>
	
	Note: Genotype columns of data exported from BeadStudio end in ".GType". This trailing ".GType" is automatically removed by SNPduo.
	<br><br>
	
	6. Proceed with using the SNPduoWeb tool as described <a href="#SNPduo">below</a>
	</p>
</blockquote>

<p>
<a href="#top">top</a>
</p>

<hr>

<!-- HapMap data analysis -->

<p>
<a name="HapMap"><strong>HapMap - Downloaded Data</strong></a>
<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SNPduoWeb can analyze data downloaded from the HapMap site. These data are very high density, and therefore have large files. The HapMap data should be run on a <strong>per chromosome</strong> basis. This portion of the tutorial will cover how to download HapMap data for use with SNPduo.
</p>

<blockquote>
	<p>
	1. First go to the <a href="http://www.hapmap.org/downloads/">Downloads</a> page at the International HapMap Project website
	<br><br>
	
	2. Click the "Genotypes" link from Bulk Data listing
	</p>
	
	<blockquote>
	<p>
	<img src="images/hapmap_genotype.jpg" width="527" height="286" alt="HapMap Genotype Download">
	</p>
	</blockquote>
	
	<p>
	3. Pick the directory of the latest data for the build you wish to use
	</p>
	
	<blockquote>
	<p>
	<img src="images/hapmap_build.jpg" width="426" height="168" alt="HapMap Build Selection">
	</p>
	</blockquote>
	
	<p>
	4. Pick the "non-redundant" folder of SNP data (you may or may not have to choose a strand first)
	<br><br>
	
	5. You should now see a list of files for each chromosome in each population. Select the chromosome and population you are interested in and click to download the file. The download will be a gzipped text file. On Linux and Mac systems the shell command "gunzip <i>filename</i>" should extract the text file from the archive. On Windows utility such as <a href="http://www.winzip.com">WinZip</a> or <a href="http://www.rarlab.com">WinRAR</a> can extract the text file from the archive.
	<br><br>
	
	6. Proceed with using the SNPduoWeb tool as described <a href="#SNPduo">below</a>
	</p>
</blockquote>

<p>
<a href="#top">top</a>
</p>

<hr>

<!-- Custom data format analysis -->
<p>
<a name="Custom"><strong>Custom Data Format</strong></a>
<br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The "Custom" data format allows for the analysis of data from any source, provided it follows the appropriate guidelines.</p>

<blockquote>
	<p>
	1. Chromosome column must have the name "Chromosome"
	<br><br>
	
	2. Physical Position column must have the name "Physical Position" or "Physical.Position"
	<br><br>
	
	3. Lines that should be skipped must start with a "#" character
	<br><br>
	
	4. Lines containing only whitespace (tabs, spaces, newline characters) are skipped
	<br><br>
	
	5. The only limitation on the number of SNPs is the maximum filesize upload limit
	</p>
</blockquote>

<p>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To use the Custom format be sure to select "Custom" from the SNPduoWeb Data Type dropdown list. Detailed instructions for running SNPduoWeb are provided <a href="#SNPduo">below</a>.

<br><br>
<a href="#top">top</a>
</p>

<hr>

<!-- SNPduoWeb explanation -->

<p>
<a name="SNPduo"><strong>Running SNPduo</strong></a>
<br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The previous sections of this tutorial describe how to properly export data for use with SNPduo. This section will cover the use of the tool itself.
<br><br>

<img src="images/duo_file.jpg" alt="SNPduoWeb Specify File">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The first step in using SNPduoWeb is to specify the file you want to analyze. By following the previous tutorials you should have a file exported in the proper format for your data type. Use the "Browse" button to locate this file.
<br><br>

<img src="images/duo_source.jpg" alt="SNPduoWeb Select Data Source">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The second step is to select the source of your data. Choose one of the options in the drop-down box. For custom formatted data be sure you have followed the rules described previously for formatting this type of input.
<br><br>

<img src="images/duo_datacolumns.jpg" alt="SNPduoWeb Data Columns">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The third and fourth steps tell SNPduoWeb where the genotype data is located. The order of who is individual one and who is individual two is not important. In each box enter the number of the column where the genotype data is located. For instance, if the headers of your input were "Chromosome     Physical Position     Individual 1     Individual 2", you would enter 3 and 4 in the box for individual 1 and individual 2, respectively.
<br><br>

<img src="images/duo_chrom.jpg" alt="SNPduoWeb Select Chromosome">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Step five specifies the chromosome of interest. There are several ways to proceed with this option.

<ol class="lowerroman">
<li>Select an individual chromosome from the choices of chromosomes 1-22, X, Y, and M<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This option will provide summary data and plot for the individual chromosome specified.</li>
<li>Select "Genome"<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This option will provide summary data and plot on one graph the entire genome</li>
<li>Select "Genome by Chromosome"<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Picking this option is the same thing as doing option "i" for every chromosome. The data is provided in a *.zip file containing the plots for each individual chromosome and the summary data for the genome</li>
</ol>

<p>
<img src="images/duo_postscript.jpg" alt="SNPduoWeb Output Page Size">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Steps six and seven allow you to specify the size of the postscript file the program generates. The default is standard letter size. Postscripts are landscape format so the width should be the widest dimension. Common paper sizes in inches are listed below:

<br>
</p>

<table border="1" cellspacing="3" cellpadding="6">
<tr align="center"><td>Letter</td><td>8.5x11</td></tr>

<tr align="center"><td>Legal</td><td>8.5x14</td></tr>

<tr align="center"><td>Government Letter</td><td>8.5x10.5</td></tr>

<tr align="center"><td>Government Legal</td><td>8.5x13</td></tr>

<tr align="center"><td>Ledger</td><td>11x17</td></tr>
</table>

<p>
<img src="images/duo_build.jpg" alt="SNPduoWeb Build Information">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The positions and extent of centromeres, telomeres, and other heterochromatin can change between genome builds. Step 8 allows you to specify which build the data is based on so the ideogram for that build is displayed correctly. Affymetrix CNAT 4.0 will be based on Build 35. Illumina BeadStudio data is based on Build 35. If you are using HapMap, select the build appropriate for the data that you downloaded. Custom data could be any build, and the user is responsible for deciding which to use.
<br><br>

<img src="images/duo_makejpeg.jpg" alt="SNPduoWeb make JPEG option">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;In step 9 you can specify whether the software should generate JPEG files of your data. The default is to create JPEGs. These JPEG files will displayed in your web browser and will be available for download. The advantage of making JPEGs is to be able to see the data in the browser without having to download the postscript and open it in a vector graphics program. The disadvantage is that making an JPEG from a postscript is a slow process and may significantly slow down the performance of the program. 
<br><br>

<img src="images/duo_runmode.jpg" alt="SNPduoWeb Normal or Batch mode options">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SNPduoWeb can be run to analyze a single pair in Normal Mode or all comparisons for a list of individuals in Batch Mode. The default mode in step 10 is Normal Mode. If you wish to run all comparisons for several individuals, select Batch Mode. Then specify the columns of the individuals you wish to compare in the "Individual 1" box, separated by commas. For example, if you want to perform all comparisons between columns 4-6, enter "4,5,6" in the Individual 1 box. Individual 2 can be left blank (and any input there will be ignored in Batch Mode).
<br><br>

<img src="images/duo_submit.jpg" alt="SNPduoWeb Submit data">
<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The final step in running SNPduoWeb is to double check the form choices you have made and to submit the data. Pressing the submit button uploads your input file to the SNPduoWeb server and triggers the supporting scripts to analyze the data using the parameters you chose. After processing a new page will display, showing a jpeg version of your IBS plot, along with links to the summary file, wig file for the UCSC Genome Browswer, jpeg image, postscript file, and a zip file containing all of the previous in one file.
</p>

<p>
<a name="Tips"><strong>Tips and Hints</strong>
<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SNPduoWeb is designed to ignore lines that start with a # character (commented lines). Any line that contains this character is automatically ignored. Also, in any context OTHER than commenting out a line the # sign is a forbidden character and will automatically be replaced by alternative characters.
<br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If you need to manually edit any header information of large text files on a Windows PC, we suggest using a more advanced text/code editor than notepad, such as <a href="http://www.crimsoneditor.com">Crimson Editor</a> or <a href="http://notepad-plus.sourceforge.net">Notepad++</a>. Mac and Linux text editors such as vi, emacs, and xemacs can usually open and manipulate large files with little problem.
<br><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If performing a "Genome" plot, it may be useful to keep the height at approximately normal landscape page size (approximately 8.5") and set the width to very long (for example 50). This will spread the plot out enough to more easily see patterns in your data.
<br><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The platform specific plotting functions are designed to work if you follow the steps outlined in this tutorial. If the data is not obtained by following those steps it should be considered Custom Data and formatted in an appropriate manner.
<br><br>
<a href="#top">top</a>
</p>
<!-- ################################  A navigation footer not needed for distribution code  ################################ -->

<table width="100%" border="0" cellpadding="6" cellspacing="3">
<tr class="rowcommon">
<td class="cell1 navcell1"><a href="index.html" class="navtext">Back</a></td>
<td class="cell2 navcell2">Return to the SNPduoWeb homepage</td>
</tr>

<tr class="rowcommon">
<td class="cell1 navcell1"><a href="/index.html" class="navtext">Home</a></td>
<td class="cell2 navcell2">Return to the Pevsner Lab Homepage</td>
</tr>

</table>

<p>
<img src="images/valid-html401-blue.png" alt="Valid HTML 4.01 Strict!" height="31" width="88">
<img src="images/vcss-blue.png" alt="Valid CSS!">
</p>

</body>
</html>